src/edata.cpp

```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
+            this->observedData.clear();
+        } else {
+            throw AmiException("Input observedData did not match dimensions nt (%i) x nytrue (%i), was %i", nt(), nytrue, observedData.size());
+        }
     }
 }
     
-void ExpData::setObservedData(const double *observedData) {
-    for (int imy = 0; imy < nytrue * nt; ++imy) {
-        my.at(imy) = static_cast<const realtype>(observedData[imy]);
+void ExpData::setObservedData(const std::vector<realtype> &observedData, int iy) {
+    if (observedData.size() == (unsigned) nt()) {
+        for (int it = 0; it < nt(); ++it)
+            this->observedData.at(iy + it*nytrue) = observedData.at(iy);
+    } else {
+        throw AmiException("Input observedData did not match dimensions nt (%i), was %i", nt(), observedData.size());
     }
 }
+    
+bool ExpData::isSetObservedData(int it, int iy) const {
+    if (!observedData.empty()) {
+        return !isNaN(observedData.at(it * nytrue + iy));
+    } else {
+        return false;
+    }
+}
+    
+std::vector<realtype> ExpData::getObservedData() const {
+    return observedData;
+}
+    
+const realtype *ExpData::getObservedDataPtr(int it) const {
+    if (!observedData.empty())
+        return &observedData.at(it*nytrue);
+    else
+        return nullptr;
+}
 
-void ExpData::setObservedDataStdDev(const double *observedDataStdDev) {
-    for (int imy = 0; imy < nytrue * nt; ++imy) {
-        sigmay.at(imy) = static_cast<const realtype>(observedDataStdDev[imy]);
+void ExpData::setObservedDataStdDev(const std::vector<realtype> &observedDataStdDev) {
+    if (observedDataStdDev.size() == (unsigned) nt()*nytrue) {
+        this->observedDataStdDev = observedDataStdDev;
+    } else {
+        if (observedDataStdDev.size() == 0) {
+            this->observedDataStdDev.clear();
+        } else {
+            throw AmiException("Input observedDataStdDev did not match dimensions nt (%i) x nytrue (%i), was %i", nt(), nytrue, observedDataStdDev.size());
+        }
     }
 }
+    
+void ExpData::setObservedDataStdDev(const realtype StdDev) {
+    std::fill(observedDataStdDev.begin() ,observedDataStdDev.end(), StdDev);
+}
+    
+void ExpData::setObservedDataStdDev(const std::vector<realtype> &observedDataStdDev, int iy) {
+    if (observedDataStdDev.size() == (unsigned) nt()) {
+        for (int it = 0; it < nt(); ++it)
+            this->observedDataStdDev.at(iy + it*nytrue) = observedDataStdDev.at(iy);
+    } else {
+        throw AmiException("Input observedDataStdDev did not match dimensions nt (%i), was %i", nt(), observedDataStdDev.size());
+    }
+}
+    
+void ExpData::setObservedDataStdDev(const realtype StdDev, int iy) {
+    for (int it = 0; it < nt(); ++it)
+        observedDataStdDev.at(iy + it*nytrue) = StdDev;
+}
+    
+bool ExpData::isSetObservedDataStdDev(int it, int iy) const {
+    if (!observedDataStdDev.empty()) {
+        return !isNaN(observedDataStdDev.at(it * nytrue + iy));
+    } else {
+        return false;
```

Comment:
`return !observedDataStdDev.empty() && !isNaN(observedDataStdDev.at(it * nytrue + iy));`

Reply:
calling `!observedDataStdDev.empty()` first that we do not perform out of bound memory access on `observedDataStdDev`.

Reply:
oh, neat